### PR TITLE
Memory efficient inverse pass

### PR DESCRIPTION
**Pull request by tychovdo**
I am training an invertible network in two directions (so also backpropagating the error after doing an inverse pass).

However, the current memcnn implementation only supports a memory efficient forward pass. So, I suggest adding inverse versions of "implementation 0" and "implementation 1" for memory-efficient inverse training. In this pull request a naive pass becomes "implementation 2".

Would be useful to add such functionality (after maybe proper unit testing!).

### CODE REVIEW

**Comment by silvandeleemput** [comment]
Hi, thank you for your contribution, a memory efficient inverse pass is very welcome and on quick inspection it appears that the implementation looks good! I would like to write the tests later this week to perform the merge.

I see you are working on the addition of an Affine coupling. I am curious about your rationale for splitting it up like this, maybe I can help out somewhere.

**Comment by tychovdo** [comment]
Nice to hear. I also added unit tests for the inverse pass.

And, I am indeed working on _affine coupling._ Please feel free to take a look and comment on the implementation. I splitted up affine and additive coupling in two separate files. The implementation seems to work (although the 'affine vs normal' unit tests are still missing).

**Comment by silvandeleemput** [comment]
Alright after taking a closer look I took the liberty to start from my old unittest and modify it to the proposed changes. Now it appears that the Additive coupling is working properly, but the Affine coupling isn't working yet (failing some unittests). You can run the updated unittests to pinpoint the errors. Here is a traceback of what I get in my environment:

```
======================================================================
FAIL: test_normal_vs_revblock (__main__.ReversibleOperationsTestCase)
ReversibleBlock test if similar gradients and weights results are obtained after similar training
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/synology/neuro/sil/tycho/memcnn/models/tests/test_revop.py", line 202, in test_normal_vs_revblock
    self.assertTrue(np.allclose(c1.weight.data.numpy(), c1_2.weight.data.numpy()))
AssertionError: False is not true

======================================================================
FAIL: test_reversible_block_fwd_bwd (__main__.ReversibleOperationsTestCase)
ReversibleBlock test of the memory saving forward and backward passes
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/synology/neuro/sil/tycho/memcnn/models/tests/test_revop.py", line 80, in test_reversible_block_fwd_bwd
    self.assertTrue(np.allclose(X.data.numpy(), data, atol=1e-06))
AssertionError: False is not true

----------------------------------------------------------------------
```

For testing purposes I would:
* disable: test_memory_saving (because it takes a while to compute)
* focus on test_normal_vs_revblock: it tests the torch.autograd.Function implementation versus normal autograd differentiation and should yield approximately similar weights and gradients. (however this currently fails on the affine coupling)
* put print statements in the inner loop to get the state of: coupling, implementation_fwd, etc...

I saw in your code that you were unsure whether to add or subtract the computed gradients on the Affine transform, you might want to flip those and run the unittests again, (I haven't tried this yet).  

If we can resolve this issue I would like to finalize the pull request.

File: memcnn/models/affine.py [hunk 1]
```diff
@@ -0,0 +1,563 @@
...
+
+            # calculate the final gradients for the weights and inputs
+            dd = torch.autograd.grad(x1_, (z1_stop,) + tuple(Fm.parameters()), x1_grad)
+            z1_grad = dd[0] + x2_grad  # + or - ?
+            FWgrads = dd[1:]
+
+            dd = torch.autograd.grad(x2_, (y2_stop, y1_stop) + tuple(Gm.parameters()), z1_grad, retain_graph=False)
+
+            GWgrads = dd[2:]
+            y1_grad = dd[1] + x1_grad  # + or - ?
```
Comment (silvandeleemput) [inline memcnn/models/affine.py]: please try this against the new unittests also for the other Affine implementations
```suggestion
            y1_grad = dd[1] - x1_grad 
```

File: memcnn/models/affine.py [hunk 2]
```diff
@@ -0,0 +1,563 @@
...
+            y.set_(torch.cat([y1, y2], dim=1).contiguous())
+
+            # compute outputs building a sub-graph
+            z1 = (y2_stop - G_y12) / G_y11
+            x1_ = (y1_stop - F_z12) / F_z11
+            x2_ = z1
+
+            # calculate the final gradients for the weights and inputs
+            dd = torch.autograd.grad(x1_, (z1_stop,) + tuple(Fm.parameters()), x1_grad)
+            z1_grad = dd[0] + x2_grad  # + or - ?
```
Comment (silvandeleemput) [inline memcnn/models/affine.py]: please try this against the new unittests also for the other Affine implementations
```suggestion
            z1_grad = dd[0] - x2_grad 
```

File: memcnn/models/additive.py [hunk 3]
```diff
@@ -0,0 +1,515 @@
...
+            y.set_(torch.cat([y1, y2], dim=1).contiguous()).detach()
+
+            # compute outputs building a sub-graph
+            z1 = y2_stop - G_y1
+            x1 = y1_stop - F_z1
+            x2 = z1
+
+            # calculate the final gradients for the weights and inputs
+            dd = torch.autograd.grad(x1, (z1_stop,) + tuple(Fm.parameters()), x1_grad)
+            z1_grad = dd[0] + x2_grad # + or - ?
```
Comment (silvandeleemput) [inline memcnn/models/additive.py]: Addition seems to be ok here.

File: memcnn/models/additive.py [hunk 4]
```diff
@@ -0,0 +1,515 @@
...
+            FWgrads = dd[1:]
+
+            dd = torch.autograd.grad(x2, (y2_stop, y1_stop) + tuple(Gm.parameters()), z1_grad, retain_graph=False)
+
+            GWgrads = dd[2:]
+            y1_grad = dd[1] + x1_grad # + or - ?
```
Comment (silvandeleemput) [inline memcnn/models/additive.py]: Addition seems to be ok here.

**Comment by robintibor** [comment]
@tychovdo have you ever used your affine invertible version? @silvandeleemput would it still be interesting to finish this stuff here? I may have a try.

**Comment by silvandeleemput** [comment]
@robintibor yes, I am still interested in finishing the merge. I just did some work on it, so at this moment the additive coupling works as intended and is covered by the unit tests. However, the affine coupling does not pass the unit tests yet, hence they have been disabled for now. 

Ideally, I would like to fix the affine coupling with the merger as well. But the merger could be completed with only the additive coupling for now. If you want, you could look into the affine coupling from the latest branch at: [https://github.com/tychovdo/memcnn](https://github.com/tychovdo/memcnn).

**Comment by robintibor** [comment]
Which pytorch version are you using? I tried to run the code in Pytorch 1.1 and it seemed that the overall approach doesn't work anymore? Or did I do something wrong? Like any call such as
`x.data.set_(...)` leads to 
`RuntimeError: set_storage is not allowed on Tensor created from .data or .detach()`
I tried to understand the autograd/backward/forward mechanics in PyTorch and as far as I understood, the whole mechanism seems designed to prevent any in-place change to a tensor stored via   `save_for_backward`? I started to write my own hacky workaround because of that. Do you have some clean solution that still runs in PyTorch 1.1?

**Comment by silvandeleemput** [comment]
MemCNN was designed and tested for PyTorch versions v0.3 and v0.4 and for Python 2.7. I haven't tested for PyTorch v1.1 yet. However, I have successfully run the unit tests with PyTorch v1.0.1.post2 and Python 3.0. 

I have some code ported tor PyTorch v1.0, but not yet for PyTorch v1.1. Later this week I might have some time to look at it and hopefully provide a clean PyTorch v1.1 solution.

**Comment by silvandeleemput** [comment]
@robintibor  I have added support for PyTorch versions v1.0.1 and v1.1. 
Pytorch v1.1 indeed required a new way to enforce the memory saving, which was achieved through .storage().resize(0), which is backward compatible.

All unittests should now pass and I will now merge this pull request with master. 
I'll make a new issue for fixing the affine transform.

**Comment by silvandeleemput** [comment]
@robintibor the new issue can be found here: #14

**Comment by robintibor** [comment]
cool @silvandeleemput great. Do you have some method of verifying that the memory on CPU and GPU is properly freed? Just wondering since I don't know what pytorch internally does there, especially if tensor shape and tensor storage are not in sync anymore.

**Comment by silvandeleemput** [comment]
@robintibor Let's continue the discussion here: #14

**Pull request merged**